WordPress/Classes/Services/NotificationSyncService.swift

```diff
@@ -0,0 +1,259 @@
+    ///
+    func updateLocalNotes(with remoteNotes: [RemoteNotification], completion: (Void -> Void)) {
+        let derivedContext = contextManager.newDerivedContext()
+        let helper = CoreDataHelper<Notification>(context: derivedContext)
+
+        derivedContext.performBlock {
+            for remoteNote in remoteNotes {
+                let predicate = NSPredicate(format: "(notificationId == %@)", remoteNote.notificationId)
+                let localNote = helper.firstObject(matchingPredicate: predicate) ?? helper.insertNewObject()
+
+                localNote.update(with: remoteNote)
+            }
+
+            self.contextManager.saveDerivedContext(derivedContext) {
+                dispatch_async(dispatch_get_main_queue()) {
+                    completion()
+                }
+            }
+        }
+    }
+
+
+    /// Deletes the collection of local notifications that cannot be found in a given collection of
+    /// remote hashes.
+    ///
+    /// - Parameter remoteHashes: Collection of remoteNotifications.
+    ///
+    func deleteLocalMissingNotes(from remoteHashes: [RemoteNotification], completion: (Void -> Void)) {
+        let derivedContext = contextManager.newDerivedContext()
+        let helper = CoreDataHelper<Notification>(context: derivedContext)
+
+        derivedContext.performBlock {
+            let remoteIds = remoteHashes.map { $0.notificationId }
+            let predicate = NSPredicate(format: "NOT (notificationId IN %@)", remoteIds)
+
+            for orphan in helper.allObjects(matchingPredicate: predicate) {
+                helper.deleteObject(orphan)
+            }
+
+            self.contextManager.saveDerivedContext(derivedContext) {
+                dispatch_async(dispatch_get_main_queue()) {
+                    completion()
+                }
+            }
+        }
+    }
+
+
+    /// Updates the Read status, of a given Notification, as specified.
+    ///
+    /// - Parameters:
+    ///     - status: New *read* value
+    ///     - noteObjectID: CoreData ObjectID
+    ///
+    func updateReadStatus(status: Bool, forNoteWithObjectID noteObjectID: NSManagedObjectID) {
+        let helper = CoreDataHelper<Notification>(context: mainContext)
+        let note = helper.loadObject(withObjectID: noteObjectID)
+        note?.read = status
+        contextManager.saveContext(mainContext)
+    }
+}
+
+
+// MARK: - Private Properties
+//
+private extension NotificationSyncService
+{
+    /// Returns the main CoredAta Context
+    ///
+    var mainContext: NSManagedObjectContext {
+        return contextManager.mainContext
+    }
+
+
+    /// Returns the WordPress.com REST API, if any
+    ///
+    var dotcomAPI: WordPressComRestApi? {
+        let service = AccountService(managedObjectContext: mainContext)
+        let account = service.defaultWordPressComAccount()
+
+        guard let dotcomAPI = account?.wordPressComRestApi where dotcomAPI.hasCredentials() else {
```

Comment:
Pretty much the standard (AFAIK). Used in Reader Services // CommentService // NotificationSettingsService.

(Please, let me know if that's actually a bad practise or so, since.. i've been using it ever since!).

Thanks!